src/gasnow_websocket.rs

```diff
@@ -0,0 +1,172 @@
+        tokio::spawn(receive_forever(
+            DEFAULT_URL.parse().unwrap(),
+            RECONNECT_INTERVAL,
+            sender,
+        ));
+        Self {
+            max_update_age,
+            receiver,
+        }
+    }
+
+    pub async fn wait_for_first_update(&mut self) {
+        while (*self.receiver.borrow_and_update()).is_none() {
+            // Unwrap because sender cannot have been dropped  while our receiver exists.
+            self.receiver.changed().await.unwrap();
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl GasPriceEstimating for GasNowWebSocketGasStation {
+    async fn estimate_with_limits(
+        &self,
+        gas_limit: f64,
+        time_limit: std::time::Duration,
+    ) -> Result<f64> {
+        if let Some((instant, response)) = *self.receiver.borrow() {
+            ensure!(
+                instant.elapsed() <= self.max_update_age,
+                "last update more than {} s in the past",
+                self.max_update_age.as_secs()
+            );
+            gasnow::estimate_with_limits(gas_limit, time_limit, &response)
+        } else {
+            bail!("did not receive first update yet");
+        }
+    }
+}
+
+#[derive(serde::Deserialize)]
+#[serde(untagged)]
+enum JsonMessage {
+    Update { data: ResponseData },
+    Other(Value),
+}
+
+/// Exits when all receivers have been dropped.
+/// Automatically reconnects the websocket.
+async fn receive_forever(
+    api: Url,
+    reconnect_interval: Duration,
+    sender: watch::Sender<Option<(Instant, ResponseData)>>,
+) {
+    let work = async {
+        loop {
+            connect_and_receive_until_error(&api, &sender).await;
+            tokio::time::sleep(reconnect_interval).await;
+        }
+    };
+    let is_closed = sender.closed();
+    futures::pin_mut!(is_closed);
+    futures::pin_mut!(work);
+    futures::future::select(work, is_closed).await;
+    tracing::debug!("exiting because all receivers have been dropped");
+}
+
+/// Returns on first error.
+async fn connect_and_receive_until_error(
+    api: &Url,
+    sender: &watch::Sender<Option<(Instant, ResponseData)>>,
+) {
+    let (mut stream, _) = match tokio_tungstenite::connect_async(api).await {
+        Ok(result) => result,
+        Err(err) => {
+            tracing::error!(?err, "websocket connect failed");
+            return;
+        }
+    };
+    while let Some(item) = stream.next().await {
+        let message = match item {
+            Ok(message) => message,
+            // It is unclear which errors exactly cause the websocket to become unusable so we stop
+            // on any.
+            Err(err) => {
+                tracing::error!(?err, "websocket failed");
+                return;
+            }
+        };
+        let json_message: Result<JsonMessage, _> = match &message {
+            Message::Text(text) => serde_json::from_str(text),
+            Message::Binary(binary) => serde_json::from_slice(binary),
```

Comment:
Do they actually send binary messages over the WS? Should this maybe be an error?

Reply:
In practice they don't but they also don't specify what messages they sent. Actually their whole documentation is just a link to some example javascript. Since it is just a couple of lines to support binary I think we can keep it for future proofing.